File: Clojure/zoo.clj
(defmulti info (fn [x] (x "kind")))

(defmethod info :dog [x] (str (x "nickname") " is a dog."))
(defmethod info :cat [x] (str (x "nickname") " is a cat."))

(def zoo (list
    {"kind" :dog "nickname" "Charlie"}
    {"kind" :dog "nickname" "Milo"}
    {"kind" :cat "nickname" "Gracie"}
    {"kind" :cat "nickname" "Molly"}
))

(doseq [x zoo]
    (println (info x)))

File: Nim/zoo.nim
type
  Animal = object of RootObj

  Cat = ref object of Animal
  Dog = ref object of Animal

  zoo_t = array[0..3, ref Animal]

proc info(x: ref Animal): string =
  if x of Cat:   return "Cat"
  elif x of Dog: return "Dog"

var zoo: zoo_t = [Cat(), Dog(), Cat(), Dog()]

for _, x in zoo:
  echo info(x)
